### PR TITLE
Uar 1595 new individual bo nocs

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-1595

### Change description

Added three new nature of control fields for the beneficial owner individual including a new NOC jurisdiction type based on changes made in the overseas web. Added validation for these fields.

As with the web, this has been feature flagged.

### Work checklist

- [x] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/BeneficialOwnerIndividualDao.java
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/BeneficialOwnerIndividualDao.java:52]: a couple of the new fields have a double 'NatureOf' in their names

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java
Comment (lduranteau) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java:17]: It would be preferable to keep the parameters relating to business logic first and keep errors and logginContext in last place

**Comment by lduranteau** [comment]
There is 1 code smell

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:80]: got the double NatureOf in this class as well

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java:35]: LAnd should have a lowercase a

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java:19]: isPropertyAndLAndNocEnabled needs lowercase a in LAnd

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidatorTest.java
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidatorTest.java:355]: can the test name be changed - i don't understand this one (or the previous one before the changes)
I think it just needs the No removing
does testErrorReportedWhenNatureOfControlValuesAreAllNullWhenNewNocsFlagIsFalse sound better?

Comment (markpit) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidatorTest.java:355]: i think the 'No" removal from the name applies to a few in this class.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/BeneficialOwnerIndividualDao.java
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/BeneficialOwnerIndividualDao.java]: So good I had to say twice, but restrained it back to once now.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java]: Rearranged parameters and arguments.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java]: As above.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java]: Altered the case.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/NatureOfControlValidators.java]: As above.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidatorTest.java
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidatorTest.java]: Changed names on 4 tests.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by mattch1** [comment]
A fourth nature of control will need to be added.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerGovernmentOrPublicAuthorityValidator.java
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerGovernmentOrPublicAuthorityValidator.java:117]: This is temporary as this will change when we get to work on these bo types. Can use a wrapper if preferred.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerGovernmentOrPublicAuthorityValidator.java
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerGovernmentOrPublicAuthorityValidator.java:117]: We could have an overloaded method in the `NatureOfControlValidators` class - one that takes two lists and another that takes just the one. I'm not a big fan of passing `nulls` in to methods (unless method is private).

File: src/test/resources/overseas_entity_v_1_0.json
Comment (mwestacott) [inline src/test/resources/overseas_entity_v_1_0.json:113]: We shouldn't be updating older versions of the test JSON files. Only the most recent one.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java
Comment (lduranteau) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:29]: To keep the names consistent, it might be preferable to call it TRUSTEES_CONTROL_NATURE_OF_CONTROL_TYPES_FIELD = trustees_control_nature_of_control_types

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java:100]: I still think this should come out of the `else` block because of the difference with the web implementation. I think it might effect how the Update journey behaves if the flag is switched on. But... guess that if the flag _is_ switched on it's acceptable that Update journey is broken as this (NOCs) is all a work in progress. So will leave it up to you @mattch1 what you do here.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:29]: But this relates to a trust NOT a trustee so I think it's correct. If it is wrong, web code would have to change too, as it defines `trust_control_nature_of_control_types`.

**Review by mwestacott — COMMENTED** [review]
LGTM

Note that you'll need to do a PR in the `ecs-service-configs-dev` repo to add the new feature flag (FEATURE_FLAG_ENABLE_PROPERTY_OR_LAND_OWNER_NOC_30082024).

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:29]: The one at line 26 is trustee not trust
And it is control over trustee not trust, look at the ticket https://companieshouse.atlassian.net/browse/UAR-1595

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:29]: I think the names have been decided upon based on the NOC headings. Heading for this one is "The beneficial owner has the right to exercise, or actually exercises, significant control or influence over the activities of a trust" (which doesn't mention 'trustee'). The heading related to line 26 is "The beneficial owner is a trustee of a trust and" so guessing why that (long ago) was named TRUSTEES_NATURE_OF_CONTROL_TYPES_FIELD.

Personally, I'm happy with this naming. Any other opinions?

Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerIndividualDto.java:29]: Mapping will change in chips-filing-consumer

**Review by lduranteau — APPROVED** [review]

**Comment by mwestacott** [comment]
Before merging I suggest you test with the web 'main' branch and your SDK branch. Check that you can get to the confirmation screen after selecting some of the new BO NOCs.

**Pull request merged**
